### ISSUE TITLE
What does Breaking Changes mean?

### ISSUE THREAD

**Issue by Atry** [hhvm/hhvm.com#187]
We commonly include a Breaking Changes in release notes. What does it actually mean? 

I used to think it is similar to breakages in binary backward compatibility in C++ or Java. However, we don't have a binary release form of a library.

If it means breakage at the source level backward compatibility, it could be too sensitive to be broken, and we might have missed many breaking change. 

For example, we used to assume changing a function from the `defaults` capabilities to pure capability is a backward compatible change. However the assumption is not true if we take type checker into account:

Given a library including the following functions:
``` hack
function method1(): void {
}
```
When changing it to
``` hack
function method1()[]: void {
}
```
 
The following user code, which could type check originally, will be failed to type check any more:

``` hack
function method2(): void {
}

function infer<T>(
  (function()[T]: void) $_,
): (function ((function()[T]: void)):void) {
  return $_ ==> {};
}

function broken(): void {
  $f = infer(method1<>);


  // Typing[4403] Invalid argumentHack(4403)
  // Expected a function that requires the capability set {}
  // But got a function that requires the capability set {AccessGlobals, IO, ImplicitPolicyLocal, RxLocal, System, WriteProperty}
  $f(method2<>); 
}
```

Similarly, changing the return type of a function into a more specific type is also backward incompatible at source level, even though similar changes are usually considered binary backward compatible in other languages.

**Comment by Atry** [comment]
(copy from internal post)

The goal of keeping binary backward compatibility in other language is to allow for upgrading a transitive dependencies. For example, given application A depending on library B depending on library C, if library B is updating to a binary backward compatible and source level backward incompatible version, application A must be migrated upon the new version to pass type check. However, if library C is updating to a binary backward compatible and source level backward incompatible version, application A does not have to migrate upon the new version.
In Hack, the maintainer of the application A cannot simply update library C, unless the developer of library B migrate first.
Something similar to binary backward compatibility in Hack is "runtime backward compatibility", if we could exclude vendor directory from type checking.

**Comment by Atry** [comment]
I think "runtime backward compatibility" is more realistic, because currently all inferred type parameters do not change runtime behavior. Reified type parameters do change runtime behavior. Fortunately it cannot be inferred.

**Comment by lexidor** [comment]
I'd go with variance rules when documenting signature changes. Non-final methods are invariant on their parameters, return type, and coeffects. All final or free standing functions are contravariant on their parameters, covariant on their return type, and contravariant on their coeffects.

Yes, there are indeed ways the typechecker can emit errors where it previously did not. These callsites can be restored by explicitly typing out the previously inferred types. If that type is not denotable, maybe mention it in the blog.

**Comment by Atry** [comment]
> Yes, there are indeed ways the typechecker can emit errors where it previously did not. These callsites can be restored by explicitly typing out the previously inferred types. If that type is not denotable, maybe mention it in the blog.

It is trivial to fix errors like this in a monorepo if all the source files of all dependencies are in the same monorepo. However, it is not easy to fix an even trivial typing error in `composer` based projects where errors are commonly found in the `vendor` directory.

**Comment by lexidor** [comment]
Would you happen to be able to give an example where project B and project C are both independently compatible with hhvm version (next). Project A depends on B, which depends on C. The error only shows up in project A, but not in a source file controlled by A? From my understanding, if C typecheck and B typechecks, wouldn't all new type errors have their primary error location in A?

**Comment by Atry** [comment]
Yes. The problem is not only about changes introduced by HHVM, but also changes introduced by third party libraries, for users who don't use a monorepo.
